### PR TITLE
Better handle discrete data bins

### LINKED ISSUE #8375: Barchart label not always correct
See number of bone scans in https://genie-private.cbioportal.org/study/summary?id=nsclc_genie_bpc

Screen Shot 2021-02-25 at 9 17 58 AM

The data is mostly 0 and sometimes positive but never negative. Either the labels should be 0 and >0, but in this case since there are not that many ticks, maybe it makes sense to show all ticks?

Same for:

image

Prolly better to show all ticks

### PR DESCRIPTION
**Pull request by onursumer**
Fix #8375

# Checks
- [ ] Runs on heroku
- [ ] Has tests or has a separate issue that describes the types of test that should be created. If no test is included it should explicitly be mentioned in the PR why there is no test.
- [ ] The commit log is comprehensible. It follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Is this PR adding logic based on one or more **clinical** attributes? If yes, please make sure validation for this attribute is also present in the data validation / data loading layers (in backend repo) and documented in [File-Formats Clinical data section](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#clinical-data)!

### CODE REVIEW

File: web/src/test/java/org/cbioportal/web/util/DataBinnerTest.java [hunk 1]
```diff
@@ -48,608 +48,10 @@
 
     @Before
     public void setup() {
-        mockData = initMockData();
+        mockData = DataBinnerMocker.mockData();
         MockitoAnnotations.initMocks(this);
     }
 
-    public Map<String, String[]> initMockData() {
```
Comment (onursumer) [inline web/src/test/java/org/cbioportal/web/util/DataBinnerTest.java:55]: Cannot add anymore data into this method due to a compiler limit (https://stackoverflow.com/a/2407930). Moved mock data into `DataBinnerMocker.java` instead.

File: web/src/main/java/org/cbioportal/web/util/DataBinHelper.java [hunk 2]
```diff
@@ -332,4 +332,71 @@ public boolean isAgeAttribute(String attributeId) {
...
+            }
+            // else keep the bin as is
+            else {
+                distinctBins.add(bin);
+            }
+        }
+        
+        // all bins except the outlier bins has to be distinct,
+        // otherwise return the original input bins (no conversion)
+        if (areAllDistinctExceptOutliers(distinctBins)) {
```
Comment (kalletlak) [inline web/src/main/java/org/cbioportal/web/util/DataBinHelper.java:378]: can't we just return `distinctBins`?

**Review by kalletlak — APPROVED** [review]

File: web/src/main/java/org/cbioportal/web/util/DataBinHelper.java (hunk 2 above)
Comment (onursumer) [reply web/src/main/java/org/cbioportal/web/util/DataBinHelper.java:378]: good point. technically we can but we need to make sure that mixing distinct bins with internal bins doesn't mess up the bar chart layout in the frontend.

**Review by inodb — APPROVED** [review]
Perfect! Thank you!

**Pull request merged**